Adding stat_summary to list of geoms

maxwellfundi opened the pull request:
@africanmathsinitiative/developers this could now be tested.

Fixes #4420

Muthenya commented:
This works fine from my end. @dannyparsons this is ready for you.

dannyparsons commented on instat/ucrGeom.vb:
```diff
@@ -159,6 +159,39 @@ Public Class ucrGeom
+
+        'These are the list of parameters on the documentation, some may have more options than what we have here which am yet to find out
+        clsgeom_statsummary.AddLayerParameter("fun.data", "list", Chr(34) & "mean_cl_boot" & Chr(34), lstParameterStrings:={Chr(34) & "mean_sdl" & Chr(34), Chr(34) & "mean_cl_boot" & Chr(34)})
+        clsgeom_statsummary.AddLayerParameter("fun.y", "list", Chr(34) & "mean" & Chr(34), lstParameterStrings:={Chr(34) & "median" & Chr(34), Chr(34) & "mean" & Chr(34)})
```
Yes some common summaries in the combo box but editable as well. I think this means we now want two types of "list" parameters, editable and non-editable, which could be done in the specification of the parameter. Can you look into doing this?

maxwellfundi replied:
Yeah I can do that :-)

bethanclarke commented:
Hi @maxwellfundi, so I thought I'd test this by trying to create the following graph:
![statsummary](https://user-images.githubusercontent.com/16190039/40417250-3af205b8-5e77-11e8-9822-f90ed63f6fa3.PNG)

Unfortunately, I R-Instat produced a graph that just looked like a scatter plot. When I ran this code in R I got the following warning: 
Computation failed in `stat_summary()`: Hmisc package required for this function 

After installing Hmisc, I got the graph above. The only thing is, I can't find where Hmisc is being used, so the `Hmisc::` notation wouldn't work here.

dannyparsons commented:
It's being used inside "mean_cl_boot" so yes we won't be able to add `Hmisc::` anywhere to fix this. Once Hmisc is installed it should then work. So if we include Hmisc as a package in R-Instat this should work.

bethanclarke commented:
stat_summary(fun.y = "mean", colour = "red", size =2) and stat_summary(fun.y = "median", colour = "red", size =2) do not work

For this to work for the `fun.y ` option, we also need to include `geom = "point"` into `stat_summary ` i.e.
stat_summary(fun.y = "mean", colour = "red", size =2, geom= "point") and the same for median option

bethanclarke commented:
@maxwellfundi This is where I am getting the graphs to test: 
http://ggplot2.tidyverse.org/reference/stat_summary.html

maxwellfundi commented:
@bethanclarke thanks for testing this work. I used that documentation you were testing with to implement this geom.  We currently have no good restrictions and linking of dependent parameters(was started but not working now) and so i think people need to be more careful here. 

@dannyparsons We had the issue of `bIsDependent` can we get this to work properly to avoid these kind of issues that @bethanclarke  has raised?

dannyparsons commented:
Do we have mandatory layer parameters yet?

dannyparsons commented:
Hmisc package now added which solve the issue raised here https://github.com/africanmathsinitiative/R-Instat/pull/4600#issuecomment-391290946

dannyparsons commented:
Has any progress been made on this? I would quite like to test this out for the next version.

maxwellfundi commented:
@dannyparsons  I am now working on this. I have added. 

So i have added this an `editablelist` as the other type of list we can have for the layer parameters then added the line to make the combo box to be editable on the `ucrReceiverMetadataProperty`
```

 ElseIf clsLayerParam.strLayerParameterDataType = "editablelist" Then
                ctrActive = ucrInputCboParamValue
                ucrInputCboParamValue.SetDropDownStyleAsEditable(True)
```
I then looked at the method on the ucrinputcombo and saw there was something that needed to be implemented here. The validation 

   ```
 Public Sub SetDropDownStyleAsEditable(bAdditionsAllowed As Boolean)
        cboInput.DropDownStyle = ComboBoxStyle.DropDown
        cboInput.AutoCompleteMode = AutoCompleteMode.Append
        cboInput.AutoCompleteSource = AutoCompleteSource.ListItems
        'TODO implement validation settings for this		
        If bAdditionsAllowed Then

        Else

        End If
    End Sub
```

So I am not sure what sort of validation we need here because any of the things we need could be anything really. So what validations did you have in mind for this?

maxwellfundi commented:
In addittion, following to your question above we only have mandatory aesthetics and that actually works.

dannyparsons commented:
That looks good. But then the combo box needs to be made non editable again in the other case.

The validation part is to adjust the validation settings in these cases, which is still to be done. But it doesn't effect the combo boxes working correctly now. It's an enhancement that is yet to be done.

maxwellfundi commented:
Sounds good @dannyparsons. I will go ahead and change the other cases to be noneditable.

maxwellfundi commented:
@bethanclarke  I have found out that because of `aes(colour = factor(vs))` on the fourth graph you could not be able create it. 

You needed to have the third receiver filled in the scatterplot dialog(we also need to add the factor()) so that the can have that graph. 

Alternatively, @bethanclarke  you needed to convert vs variable to factor and you would have exactly the same graph.  So this works.

maxwellfundi commented:
@bethanclarke  you can now retest this

@africanmathsinitiative/developers This is ready for review

bethanclarke commented:
Okay I will test this today :)

maxwellfundi commented:
@bethanclarke ?

bethanclarke commented:
I have tested this and it is working well, @dannyparsons this is ready for testing, thanks.
